runners/spark/src/main/java/org/apache/beam/runners/spark/metrics/SparkMetricsContainer.java

```diff
@@ -0,0 +1,300 @@
+        MetricUpdates cumulative = container.getCumulative();
+        Iterable<CounterAggregator> counterAggregators =
+            Iterables.transform(cumulative.counterUpdates(),
+                new Function<MetricUpdate<Long>, CounterAggregator>() {
+                  @SuppressWarnings("ConstantConditions")
+                  @Nullable
+                  @Override
+                  public CounterAggregator
+                  apply(MetricUpdate<Long> update) {
+                    return new CounterAggregator(new SparkMetricKey(update.getKey()),
+                        update.getUpdate());
+                  }
+                });
+        Iterable<DistributionAggregator> distributionAggregators =
+            Iterables.transform(cumulative.distributionUpdates(),
+                new Function<MetricUpdate<DistributionData>, DistributionAggregator>() {
+                  @SuppressWarnings("ConstantConditions")
+                  @Nullable
+                  @Override
+                  public DistributionAggregator
+                  apply(MetricUpdate<DistributionData> update) {
+                    return new DistributionAggregator(new SparkMetricKey(update.getKey()),
+                        new SparkDistributionData(update.getUpdate()));
+                  }
+                });
+        aggregate(counterAggregators, distributionAggregators);
+      }
+    }
+  }
+
+  private void aggregate(Iterable<CounterAggregator> counterUpdates,
+                         Iterable<DistributionAggregator> distributionUpdates) {
+    for (CounterAggregator update : counterUpdates) {
+      MetricKey key = update.getKey();
+      CounterAggregator current = counters.get(key);
+      Long counterUpdate = update.getValue();
+      CounterAggregator newValue;
+      if (current != null) {
+        newValue = current.updated(counterUpdate);
+      } else {
+        newValue = new CounterAggregator(key, counterUpdate);
+      }
+      counters.put(new SparkMetricKey(key), newValue);
+    }
+    for (DistributionAggregator update : distributionUpdates) {
+      MetricKey key = update.getKey();
+      DistributionAggregator current = distributions.get(key);
+      DistributionData distributionUpdate = update.getValue();
+      DistributionAggregator newValue;
+      if (current != null) {
+        newValue = current.updated(distributionUpdate);
+      } else {
+        newValue = new DistributionAggregator(key, distributionUpdate);
+      }
+      distributions.put(new SparkMetricKey(key), newValue);
+    }
+  }
+
+  private static class MetricsContainerCacheLoader extends CacheLoader<String, MetricsContainer> {
+    @SuppressWarnings("NullableProblems")
+    @Override
+    public MetricsContainer load(String stepName) throws Exception {
+      return new MetricsContainer(stepName);
+    }
+  }
+
+  private static class SparkDistributionData extends DistributionData implements Serializable {
```

Comment:
AutoValue should support Serializable. Either:

1. Add `implements Serializable` to the existing DistributionData, or:
2. Use `AutoValue` to create this subclass:
```
@AutoValue
public static class SparkDistributionData extends DistributionData implements Serializable {
  public static SparkDistributionData create(...) {
    return new AutoValue_SparkDistributionData(...);
  }
}
```